maestro-ios-driver/src/main/kotlin/xcuitest/XCTestDriverClient.kt

```diff
@@ -185,62 +181,119 @@ class XCTestDriverClient(
 
         val requestBuilder = Request.Builder()
             .addHeader("Content-Type", "application/json")
-            .url(httpUrl)
+            .url(client.xctestAPIBuilder(pathSegment).build())
             .post(bodyData)
 
         return okHttpClient
             .newCall(requestBuilder.build())
-            .execute()
+            .execute().use { processResponse(it, pathSegment) }
     }
 
-    private fun executeJsonRequest(url: String, body: Any): Response {
-        val response = executeJsonRequestUNCHECKED(url, body)
-
-        if (!response.isSuccessful) {
-            val responseBodyAsString = response.body?.bytes()?.let {
-                String(it)
-            }
-            val code = response.code
-            logger.error("Request for $url failed, status code ${code}, body: $responseBodyAsString")
-            error("Request for $url failed, status code ${code}, body: $responseBodyAsString")
-        }
+    private fun executeJsonRequest(pathSegment: String): String {
+        val requestBuilder = Request.Builder()
+            .url(client.xctestAPIBuilder(pathSegment).build())
+            .get()
 
-        return response
+        return okHttpClient
+            .newCall(requestBuilder.build())
+            .execute().use { processResponse(it, pathSegment) }
     }
 
-    private inline fun <reified T: Any> executeJsonRequest(url: String, body: Any, type: KClass<T>): T {
-        val response = executeJsonRequest(url, body)
+    private fun processResponse(response: Response, url: String): String {
+        val responseBodyAsString = response.body?.bytes()?.let { bytes -> String(bytes) } ?: ""
 
-        return response.body.use { responseBody ->
-            responseBody ?: error("Missing response body for mapping to $type")
+        return if (!response.isSuccessful) {
+            val code = response.code
+            handleExceptions(code, url, responseBodyAsString)
+        } else {
+            responseBodyAsString
+        }
+    }
 
-            mapper.readValue(responseBody.bytes(), T::class.java)
-                ?: error("Response body '${String(responseBody.bytes()).take(10)}...' not mappable to $type")
+    private fun handleExceptions(
+        code: Int,
+        pathString: String,
+        responseBodyAsString: String,
+    ): String {
+        val error = mapper.readValue(responseBodyAsString, Error::class.java)
+        when {
+            code in 400..499 -> {
+                logger.error("Request for $pathString failed with bad request ${code}, body: $responseBodyAsString")
+                throw XCUITestServerError.BadRequest(
+                    "Request for $pathString failed with bad request ${code}, body: $responseBodyAsString",
+                    responseBodyAsString
+                )
+            }
+            code == 502 -> {
+                logger.error("Request for $pathString failed, because of XCUITest server got crashed/exit, body: $responseBodyAsString")
```

Comment:
Yes this is because on network layer, 502 is surfaced when we try to reestablish connection. So, its a special case. For that context I'll change 502 to use same constant variable.